File: 109.convert-sorted-list-to-binary-search-tree.py
#
# @lc app=leetcode id=109 lang=python3
#
# [109] Convert Sorted List to Binary Search Tree
#

# Time:  O(n)
# Space: O(logn)


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


# @lc code=start
class Solution:
    def sortedListToBST(self, head: ListNode) -> TreeNode:
        def helper(head: ListNode, tail: ListNode):
            if head == tail:
                return None
            slow, fast = head, head
            while fast != tail and fast.next != tail:
                slow = slow.next
                fast = fast.next.next
            node = TreeNode(slow.val)
            node.left = helper(head, slow)
            node.right = helper(slow.next, tail)
            return node

        if not head:
            return None
        return helper(head, None)


# @lc code=end

File: 109.convert-sorted-list-to-binary-search-tree.js
/*
 * @lc app=leetcode id=109 lang=javascript
 *
 * [109] Convert Sorted List to Binary Search Tree
 */

function ListNode(val, next) {
  this.val = val === undefined ? 0 : val;
  this.next = next === undefined ? null : next;
}

function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val;
  this.left = left === undefined ? null : left;
  this.right = right === undefined ? null : right;
}

// @lc code=start
/**
 * @param {ListNode} head
 * @return {TreeNode}
 */
var sortedListToBST = function (head) {
  const helper = (head, tail) => {
    if (head === tail) {
      return null;
    }
    let slow = head;
    let fast = head;
    while (fast !== tail && fast.next !== tail) {
      slow = slow.next;
      fast = fast.next.next;
    }
    const node = new TreeNode(slow.val);
    node.left = helper(head, slow);
    node.right = helper(slow.next, tail);
    return node;
  };
  return head ? helper(head, null) : null;
};
// @lc code=end
